File: src/embeddings/validate.ts
import {
  resolveEmbeddingBaseUrl,
  resolveEmbeddingDimensions,
  resolveEmbeddingModel,
} from "./openai.ts";

export interface ValidateApiKeyResult {
  valid: boolean;
  status?: number;
  message: string;
}

export async function validateEmbeddingApiKey(options: {
  apiKey: string;
  baseUrl?: string;
  model?: string;
  dimensions?: number;
}): Promise<ValidateApiKeyResult> {
  const baseUrl = (options.baseUrl ?? resolveEmbeddingBaseUrl()).replace(/\/$/, "");
  const model = options.model ?? resolveEmbeddingModel();
  const dimensions = options.dimensions ?? resolveEmbeddingDimensions(model);
  const endpoint = `${baseUrl}/embeddings`;

  const body: Record<string, unknown> = {
    model,
    input: "miru setup validation",
  };
  if (dimensions != null) {
    body.dimensions = dimensions;
  }

  let response: Response;
  try {
    response = await fetch(endpoint, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${options.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      valid: false,
      message: `Could not reach embedding API at ${baseUrl}: ${message}`,
    };
  }

  if (!response.ok) {
    const text = await response.text();
    if (response.status === 401 || response.status === 403) {
      return {
        valid: false,
        status: response.status,
        message: "Invalid API key (authentication failed).",
      };
    }
    return {
      valid: false,
      status: response.status,
      message: `Embedding API returned ${response.status}: ${text.slice(0, 200)}`,
    };
  }

  try {
    const payload = (await response.json()) as {
      data?: Array<{ embedding?: number[] }>;
    };
    const embedding = payload.data?.[0]?.embedding;
    if (!embedding || embedding.length === 0) {
      return {
        valid: false,
        status: response.status,
        message: "Embedding API returned an empty response.",
      };
    }
    if (dimensions != null && embedding.length !== dimensions) {
      return {
        valid: false,
        status: response.status,
        message: `Expected ${dimensions} embedding dimensions, got ${embedding.length}.`,
      };
    }
  } catch {
    return {
      valid: false,
      status: response.status,
      message: "Embedding API returned invalid JSON.",
    };
  }

  return { valid: true, status: response.status, message: "API key is valid." };
}

File: internal/embeddings/validate.go
package embeddings

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type ValidateResult struct {
	Valid   bool
	Status  int
	Message string
}

func ValidateAPIKey(apiKey string, baseURL string, model string, dimensions *int) ValidateResult {
	if baseURL == "" {
		baseURL = ResolveEmbeddingBaseURL()
	}
	baseURL = strings.TrimRight(baseURL, "/")
	if model == "" {
		model = ResolveEmbeddingModel()
	}
	if dimensions == nil {
		if value, ok := ResolveEmbeddingDimensions(model); ok {
			dimensions = &value
		}
	}
	body := map[string]any{
		"model": model,
		"input": "miru setup validation",
	}
	if dimensions != nil {
		body["dimensions"] = *dimensions
	}
	data, err := json.Marshal(body)
	if err != nil {
		return ValidateResult{Valid: false, Message: err.Error()}
	}
	req, err := http.NewRequest(http.MethodPost, baseURL+"/embeddings", bytes.NewReader(data))
	if err != nil {
		return ValidateResult{Valid: false, Message: err.Error()}
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return ValidateResult{Valid: false, Message: fmt.Sprintf("Could not reach embedding API at %s: %s", baseURL, err.Error())}
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
			return ValidateResult{Valid: false, Status: resp.StatusCode, Message: "Invalid API key (authentication failed)."}
		}
		text := string(raw)
		if len(text) > 200 {
			text = text[:200]
		}
		return ValidateResult{Valid: false, Status: resp.StatusCode, Message: fmt.Sprintf("Embedding API returned %d: %s", resp.StatusCode, text)}
	}
	var payload struct {
		Data []struct {
			Embedding []float64 `json:"embedding"`
		} `json:"data"`
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return ValidateResult{Valid: false, Status: resp.StatusCode, Message: "Embedding API returned invalid JSON."}
	}
	if len(payload.Data) == 0 || len(payload.Data[0].Embedding) == 0 {
		return ValidateResult{Valid: false, Status: resp.StatusCode, Message: "Embedding API returned an empty response."}
	}
	if dimensions != nil && len(payload.Data[0].Embedding) != *dimensions {
		return ValidateResult{Valid: false, Status: resp.StatusCode, Message: fmt.Sprintf("Expected %d embedding dimensions, got %d.", *dimensions, len(payload.Data[0].Embedding))}
	}
	return ValidateResult{Valid: true, Status: resp.StatusCode, Message: "API key is valid."}
}
